### DISCUSSION TITLE
How do I fire an event that performs an `async` call and `assign`s the result to the `context`

### DISCUSSION THREAD

**Discussion by escritorio-gustavo** [statelyai/xstate#4683] [Q&A]
I'm creating a Tauri app using SolidJS and xstate. I want to read from a file with the following work flow:
1- click button
2- call tauri apis to read file
3- `await` file read operation
4- `assign` the result

I have done this in the past through some dodgy code, and I want to know whats the proper way to do this with xstate

**Comment by Andarist** [comment]
You can return data that you want to assign to context from your async job and use that to `assign` in `onDone`:
```ts
invoke: {
  src: 'asyncJobThatReturnsTheData',
  onDone: { actions: assign(({ event }) => event.output) } 
}
```

A more complete example can be found [here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgZwKYwK5gDRwGZQQgAKhIwauAhsssAOYB2cAvvmXAEQAeyMVMVJwDccAPRi4MAJ5hUyAFxwArADo1ARgBQWtJjAAKBFrhTZ8pQjY1EJ03FQA3VIxiLEZuUs4A5AKIAGgAqnKzCdizYdlQAxjDQ7sb2+OgxABYAsqgAgkFpwIz0SgREpEQUqAY00owxcAYAlHAAvAB89UhU3jD5haEsDQ1RppFaA6oxUKgCqBmxvZVJcAXAMMBUADbeeBAQnMMo-IKJdqY7EJan9hCMl8nJ-sHeAEZUUPtXIwdfV69Qd-cCo4IABrVAA+4oKAxbapTI5PIFegfSGmG4AERu4NsqPssTWN3cNDoTAMRgczlcrCabRxuPsThcMHC9PsEjgAD0APyfVExQnwCAYGBgYUtClM1RCkXClms8SSezc3mQqaYKDMRmuKXC0XMlX2Abfe6jSGmw0HUYDcJAA)

**Reply by escritorio-gustavo** [reply]
This does work, but only on the top level of the state. When I attempt to do this in an event TS starts throwing errors

**Reply by escritorio-gustavo** [reply]
The case I need the event to work is as follows:
I have a search running in a service initiated in a state called "running". The search is paginated on the app's backend. The pagination controls can be clicked at any time during the search and they must not change the state away from "running". Since the pagination happens in the backend, fetching a page is an `async` operation.

The way I have been doing this is invoking the pagination on the button's onClick event and then sending the data to the machine through an event, but this feels really hacky, as it seems like this type of logic should be contained in the machine and abstracted in some kind of "setPage" event

**Reply by Andarist** [reply]
> When I attempt to do this in an event TS starts throwing errors

Could you show me an example of what (and how) you are trying to do? It's easier to discus those things using concrete code examples.

**Reply by escritorio-gustavo** [reply]
I'm trying to do something like this
```ts
export const machine = createMachine(
  {
    id: 'UIStateMachine',
    initial: 'idle',
    context,
    states: {
      idle: {
        on: {
          run: {
            target: 'running'
          }
        }
      }
      running: {
        invoke: {
          src: 'asyncJob'
          onDone: {
            target: 'idle'
          }
        },
        on: {
          setPage: {
            invoke: 'asyncGetSelectedPage',
            onOone: {
              actions: assign({ /* ... */ })
            }
          }
        }
      }
    }
  },
  {
    services: {
      async asyncJob() {
        await something()
      },
      async asyncGetSelectedPage(ctx) {
        return await getPage(ctx.pageIndex)
      }
    },
  },
)
```

**Reply by escritorio-gustavo** [reply]
But it seems like `invoke` isn't allowed in events

**Reply by Andarist** [reply]
> But it seems like invoke isn't allowed in events

Yes, that's the case. It's an interesting proposal though.

**Reply by escritorio-gustavo** [reply]
Well, that's fine. From what I've been reading it seems that in v5 something similar to what I'm trying to achieve would be possible (someone deleted a reply using `fromCallBack` with svelte that looked interesting, though it was deleted before I managed to figure it out). That's unfortunately not feasible for me since I'm using Solid, but until `@xstate/solid` becomes compatible with v5 I'll keep using the workaround I've been using.

```tsx
const machine = createMachine(
  {
    id: 'UIStateMachine',
    initial: 'idle',
    context,
    states: {
      idle: {
        on: {
          run: {
            target: 'running'
          }
        }
      }
      running: {
        invoke: {
          src: 'asyncJob'
          onDone: {
            target: 'idle'
          }
        },
      }
    },
    on: {
      setPage: {
        onOone: {
          actions: assign({ /* ... */ })
        }
      }
    }
  },
  {
    services: {
      async asyncJob() {
        await something()
      },
    },
  },
)

function Component() {
  const [state, send] = useMachine(machine)
  return (
    <button onClick={async () => {
       const data = await getPage(ctx.pageIndex)
       send({ type: 'setPage', data })
    }}>
  )
}
```

**Reply by Andarist** [reply]
> but until @xstate/solid becomes compatible with v5 I'll keep using the workaround I've been using.

Keep an eye on [this issue](https://github.com/statelyai/xstate/issues/3925#issuecomment-1896189619)

**Reply by escritorio-gustavo** [reply]
> Keep an eye on https://github.com/statelyai/xstate/issues/3925#issuecomment-1896189619

Will do. Thank you

**Comment by escritorio-gustavo** [comment]
Now that v5 for solid is out, what would be the "proper" way to achieve this?

I want the `setPage` event to:
- receive a `pageIndex`
- fetch a page from tauri (this is an `async` operation)
- `assign` to `pagIndex` the index received by the event
- `assign` to `page` the data returned by tauri

```ts
on: {
    setPage: {
        actions: assign({
            pageIndex: ({ event }) => event.pageIndex,
            page: ({ event }) => /* fetch data at event.pageIndex */
        }),
    },
}
```
